addons/mail/models/ir_attachment.py

```diff
@@ -124,4 +124,7 @@ def _to_store(self, store: Store, /, *, fields=None, extra_fields=None):
                     if attachment.res_model != "mail.compose.message" and attachment.res_id
                     else False
                 )
+            if not self.env.user._is_internal():
+                attachment.generate_access_token()
```

Comment:
Here you are creating an access token for every attachment that goes through `_to_store` if the user is a portal user. This is not what we want. The access token must only be generated if the attachment relates to a `knowledge.article.thread`, in the context of posting a message, if and only if the user has write access to the article the thread is related to.

This should be done in `knowledge_article_thread.py:_process_attachments_for_post` (should override the function in that extension).
You should check that the user has write access to the article linked to the thread before generating the access token.
You should then extend `IrAttachment` and override `_to_store` in `knowledge` to add `access_token` in data only when it is related to `knowledge.article.thread`, and only if the user as read or write access to the related article.